docs/architecture/adr-111-nil-mempool.md

```diff
@@ -0,0 +1,239 @@
+*	`RemoveTxByKey(txKey types.TxKey) error`: returns `ErrNotFound`
+* `ReapMaxBytesMaxGas(maxBytes, maxGas int64) types.Txs`: returns `nil`
+* `ReapMaxTxs(max int) types.Txs`: returns `nil`
+*	`Lock()`: does nothing
+* `Unlock()`: does nothing
+*	`Update(...) error`: returns `nil`
+* `FlushAppConn() error`: returns `nil`
+*	`Flush()`: does nothing
+* `TxsAvailable() <-chan struct{}`: returns a closed channel
+* `EnableTxsAvailable()`: does nothing
+* `SetTxRemovedCallback(cb func(types.TxKey))`: does nothing
+* `Size() int` returns 0
+* `SizeBytes() int64` returns 0
+
+Upon startup, the `nil` mempool reactor will advertise no channels to the peer to peer layer.
+
+### Configuration
+
+We propose the following changes to the `config.toml` file:
+
+```toml
+[mempool]
+# The type of mempool for this CometBFT node to use.
+#
+# Valid types of mempools supported by CometBFT:
+# - "flood" : clist mempool with flooding gossip protocol (default)
+# - "nil"   : nil-mempool (app needs an alternative tx dissemination mechanism)
+type = "nil"
+```
+
+The config validation logic will be modified to add a new rule that rejects a configuration file if:
+
+* the mempool is set to `nil`
+* `create_empty_blocks`, in `consensus` section, is set to `false`.
+
+### Impacted Workflows
+
+* *Submitting a transaction*. Users are not to submit transactions via CometBFT's RPC.
+  `BroadcastTx*` RPC methods will fail.
+  The application running on a full node must offer an interface for users to submit new transactions.
+  It could also be a distinct node (or set of nodes) in the network.
+  These considerations are exclusively the application's concern in this approach.
+* *Time to propose a block*. The consensus reactor will call `ReapMaxBytesMaxGas` which will return a `nil` slice.
+  `RequestPrepareProposal` will thus contain no transactions.
+* *Consensus waiting for transactions to become available*. `TxsAvailable()` returns a closed channel,
+  so consensus doesn't block (potentially producing empty blocks).
+  At any rate, a configuration with the `nil` mempool and `create_empty_blocks` set to `false`
+  will be rejected at the first place.
+* *A new block is decided*.
+  * When `Update` is called, nothing is done (no decided transaction is removed).
+  * Locking and unlocking the mempool has no effect.
+
+### Impact on Current Release Plans
+
+The changes needed for this approach, are fairly simple, and the logic is clear.
+This might allow us to even deliver it as part of CometBFT `v1` (our next release)
+even without a noticeable impact on `v1`'s delivery schedule.
+
+The CometBFT team (learning from past dramatic events) usually takes a conservative approach
+for backporting changes to release branches that have already undergone a full QA cycle
+(and thus are in code-freeze mode).
+For this reason, although the limited impact of these changes would limit the risks
+of backporting to `v0.38.x` and `v0.37.x`, a careful risk/benefit evaluation will
```

Comment:
backport to `v0.38.x-experimental` ? Or as an opt-in?

Reply:
That's part of the discussion we'll need to have, and which I'd like to keep out of the scope of this ADR.